### PR TITLE
Exercise 03 -  Setup one-click local development in python.

### PR DESCRIPTION
**Pull request by aashirathore6362**
Please review the assignment, let me know if there are any improvements.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,8 @@
+FROM python:alpine
+WORKDIR /Rest-api
```
Comment (sitaramshelke) [inline Dockerfile:2]: just use lowercase everywhere possible in non-code scenario

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+COPY requirements.txt makefile student.py /Rest-api/
```
Comment (sitaramshelke) [inline Dockerfile:3]: Make a note to yourself to read and understand about docker layer catching optimization tricks to reduce the docker build time. 
This is a standard practice and you will find enough material online to read.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,8 @@
+FROM python:alpine
+WORKDIR /Rest-api
+COPY requirements.txt makefile student.py /Rest-api/ 
+RUN apk update
```
Comment (sitaramshelke) [inline Dockerfile:4]: Related to above comment, read about combining multiple `RUN` commands to speed up docker build process.

File: makefile [hunk 4]
```diff
@@ -0,0 +1,5 @@
+install-dep:
+	pip install -r requirements.txt
+run:install-dep
+	flask -A student.py run --host=0.0.0.0 --port=4000
```
Comment (sitaramshelke) [inline makefile:4]: Port should be configurable.
Apply 12 factor app methods to move it into Environment config.

File: student.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        return make_response(jsonify({'getstd': getstd.json()}), 200)
+    return make_response(jsonify({'message': 'student not found'}, 404))
+
+# Update student
+@app.route('/api/v1/student/<int:id>', methods=['PUT'])
+def update_std(id):
+    updatestd = User.query.filter_by(id=id).first()
+    if updatestd:
+      data = request.get_json()
+      updatestd.stdname = data['stdname']
```
Comment (sitaramshelke) [inline student.py:57]: Think if someone posts empty string for name and title. What should be returned?

**Review by sitaramshelke — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (aashirathore6362) [reply Dockerfile]: I've made changes to ensure lowercase everywhere in code.

File: Dockerfile (hunk 3 above)
Comment (aashirathore6362) [reply Dockerfile]: I've combined multiple run commands in dockerfile.

File: student.py (hunk 5 above)
Comment (aashirathore6362) [reply student.py]: Adding the exceptional handling in student.py file for this.

**Comment by aashirathore6362** [comment]
@sitaramshelke Kindly review the changes and let me know your inputs on the same.

File: Dockerfile [hunk 6]
```diff
@@ -7,9 +7,12 @@ RUN pip install --user -r requirements.txt
 
 #App run ->Stage2
 FROM python:alpine
-COPY --from=baseimage /root/.local /root/.local
-COPY student.py /app/
 WORKDIR /app
+COPY --from=baseimage /root/.local /root/.local
+COPY student.py runapp.sh /app/
+COPY migrations /app/migrations/
```
Comment (sitaramshelke) [inline Dockerfile:13]: This could have been a single COPY statement

**Review by sitaramshelke — CHANGES_REQUESTED** [review]
I understand you have used `MIGRATION` environment variable to differentiate between app run vs app migration.

I would have also tried in a different way. Utilizing `Makefile` and Docker `CMD` feature

So I would create a Make task which runs both `flask db migrate` and `flask db upgrade` commands. So my DB migrate container in `docker-compose.yml` would have a custom `CMD` as `make db_migrate` 
And this way I wouldn't even need the `runapp.sh` file.

**Review by sitaramshelke — APPROVED** [review]

File: Dockerfile (hunk 6 above)
Comment (aashirathore6362) [reply Dockerfile]: I've combined multiple COPY commands in single command.

**Comment by aashirathore6362** [comment]
I understand the strategy you described above, but in my view this way we should copy a makefile to inside a container and this is not a good way to do. 
make file is used to specially for test/compile the locally changes not in other env, also we run migration in one time so utilizing a makefile is not a got way.

**Comment by sitaramshelke** [comment]
> make file is used to specially for test/compile the locally changes not in other env,            

I don't think there is any such rule
and
> we run migration in one time so utilizing a makefile is not a got way.    

I don't think Makefile has anything to do with how many time a task is invoked.

**Pull request merged**